crypto/evp_extra/evp_extra_test.cc

```diff
@@ -1695,3 +1699,434 @@ TEST_P(EVPRsaPssBadKeyTest, InvalidSaltLength) {
+  EXPECT_EQ(pk_len, GetParam().public_key_len);
+  EXPECT_EQ(sk_len, GetParam().secret_key_len);
+
+  // Test the EVP_PKEY_size function.
+  int pkey_size = GetParam().public_key_len + GetParam().secret_key_len;
+  EXPECT_EQ(EVP_PKEY_size(pkey.get()), pkey_size);
+}
+
+// Helper function that:
+//   1. creates EVP_PKEY_CTX object of KEM type,
+//   2. sets the KEM parameters according to the given nid,
+//   3. generates a key pair,
+//   4. creates EVP_PKEY object from the generated key,
+//   5. creates a new context with the EVP_PKEY object and returns it.
+static bssl::UniquePtr<EVP_PKEY_CTX> setup_ctx_and_generate_key(int kem_nid) {
+
+  // Create context of KEM type.
+  bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_KEM, nullptr));
+  EXPECT_TRUE(ctx);
+
+  // Setup the context with specific KEM parameters.
+  EXPECT_TRUE(EVP_PKEY_CTX_kem_set_params(ctx.get(), kem_nid));
+
+  // Generate a key pair.
+  EVP_PKEY *raw = nullptr;
+  EXPECT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+  EXPECT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  EXPECT_TRUE(raw);
+
+  // Create PKEY from the generated raw key and a new context with it.
+  bssl::UniquePtr<EVP_PKEY> pkey(raw);
+  ctx.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+  EXPECT_TRUE(ctx);
+
+  return ctx;
+}
+
+TEST_P(PerKEMTest, Encapsulation) {
+
+  // ---- 1. Setup phase: generate a context and a key ----
+  bssl::UniquePtr<EVP_PKEY_CTX> ctx;
+  ctx = setup_ctx_and_generate_key(GetParam().nid);
+  ASSERT_TRUE(ctx);
+
+  // ---- 2. Test basic encapsulation flow ----
+  // Alloc ciphertext and shared secret with the expected lenghts.
+  size_t ct_len = GetParam().ciphertext_len;
+  size_t ss_len = GetParam().shared_secret_len;
+  std::vector<uint8_t> ct(ct_len);
+  std::vector<uint8_t> ss(ss_len);
+
+  // Encapsulate.
+  ASSERT_TRUE(EVP_PKEY_encapsulate(ctx.get(), ct.data(), &ct_len, ss.data(), &ss_len));
+
+  // Check the lengths set by encapsulate are as expected.
+  EXPECT_EQ(ct_len, GetParam().ciphertext_len);
+  EXPECT_EQ(ss_len, GetParam().shared_secret_len);
+
+  // ---- 3. Test getting the lengths only ----
+  // Reset the length variables.
+  ct_len = 0;
+  ss_len = 0;
+
+  // Get the lengths and check them.
+  ASSERT_TRUE(EVP_PKEY_encapsulate(ctx.get(), nullptr, &ct_len, nullptr, &ss_len));
+  EXPECT_EQ(ct_len, GetParam().ciphertext_len);
+  EXPECT_EQ(ss_len, GetParam().shared_secret_len);
+
+  // ---- 4. Test calling encapsulate with different lengths ----
+  // Set ct length to be less than expected -- should fail.
+  ct_len = GetParam().ciphertext_len - 1;
+  ASSERT_FALSE(EVP_PKEY_encapsulate(ctx.get(), ct.data(), &ct_len, ss.data(), &ss_len));
+
+  // Set ct length to be greater than expected -- should succeed because
+  // it's ok to provide buffer that's larger than needed.
+  ct_len = GetParam().ciphertext_len + 1;
+  ASSERT_TRUE(EVP_PKEY_encapsulate(ctx.get(), ct.data(), &ct_len, ss.data(), &ss_len));
+
```

Comment:
This test is narrowly avoid a memory corruption bug:
```
  std::vector<uint8_t> ct(ct_len);
  ct_len = GetParam().ciphertext_len + 1;
  ASSERT_TRUE(EVP_PKEY_encapsulate(ctx.get(), ct.data(), &ct_len, ss.data(), &ss_len));
```
The buffer is still only `ct_len` long but you're telling EVP_PKEY_encapsulate it can use up to `ct_len + 1` which we don't and address sanitizer is ensuring that.